controllers/machine_controller_noderef.go

```diff
@@ -163,30 +164,17 @@ func summarizeNodeConditions(node *corev1.Node) (corev1.ConditionStatus, string)
 }
 
 func (r *MachineReconciler) getNode(ctx context.Context, c client.Reader, providerID *noderefutil.ProviderID) (*corev1.Node, error) {
-	log := ctrl.LoggerFrom(ctx, "providerID", providerID)
-
 	nodeList := corev1.NodeList{}
-	for {
-		if err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue)); err != nil {
-			return nil, err
-		}
-
-		for _, node := range nodeList.Items {
-			nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
-			if err != nil {
-				log.Error(err, "Failed to parse ProviderID", "node", node.Name)
-				continue
-			}
-
-			if providerID.Equals(nodeProviderID) {
-				return &node, nil
-			}
-		}
+	if err := c.List(ctx, &nodeList, client.MatchingFields{remote.NodeProviderIDIndex: providerID.IndexKey()}); err != nil {
+		return nil, err
+	}
+	if len(nodeList.Items) == 0 {
+		return nil, ErrNodeNotFound
+	}
 
-		if nodeList.Continue == "" {
-			break
-		}
+	if len(nodeList.Items) > 1 {
+		return nil, fmt.Errorf("unexpectedly found more than one Node matching the providerID %s", providerID.String())
```

Comment:
Instead of erroring out, could we still use the for loop and validate the equality of the provider id?

We should also make to fallback to list all nodes in case the list doesn't return any items. This could save us later in case we need changes to how the providerIDs are formed.

Reply:
If we are using the same indexKey we use for the equality check i.e `cloud/id`, there's no additional value on doing the two loops you describe above right? the client indexing already had to do that loop underneath.

Reply:
If for whatever reason the index isn't registered or available, or the fake client is used this might fail

Reply:
ok I updated to fallback and loop over the whole list when no nodes are found via index.

I'd error for the case where more than 1 are found rather than overcomplicating the func by reimplementing the logic to count with no indexes. At the moment for this case we are just returning the first one, erroring seems an improvement. Please let me know if you disagree.